Add support for client indexes

jcaamano opened the pull request:
A user may specify additional indexes at runtime through the client DB
model at client initialization. Indexes are specified in sets per table.
Each set consists of columns and optionally a key within the column that
can be addressed if the column type is a map.

The main use case is performance. Running a predicate on all the rows of
a table can be costly. Introducing new indexes in the schema is not backwards
compatible or might not align with the general concept of the schema but those
additional indexes might be useful for specific uses of that schema that 
would otherwise rely on running less performant predicates frequently.

For example, with OVN-K, checking whether ~40 LBs are present in the cache
by a predicate that matches by name on a table that contains ~200K rows took 
around 3 seconds. While surely there is different ways to optimize a predicate, 
for OVN-K which gives & frequently addresses LBs by unique names would 
make sense to have them indexed, something that is not possible to do with 
the schema without breaking backward compatibility.

This can be enhanced in the future with more use cases. For example, supporting
primary and secondary client indexes where primary client indexes could be 
checked for duplicates against a transaction before sending it to the server.

This is implemented in the cache tracking client indexes as if they were
schema indexes. Indexes map to a set of uuids instead of a single uuid
where the set will consist of a single uuid for schema index but might
consist of multiple uuids for client indexes. A new `RowsByModel` method
queries the cache considering schema and client indexes. `RowsByCondition`
also makes use of the new client indexes naturally. The legacy
`RowByModel` method considers only schema indexes. Schema indexes have
priority over client indexes in general.

API wise, client indexes are leveraged through the conditional
APIs from `Where(...)` and `WhereAll(...)`.

So where you would have before something like
```
// slow predicate run on all the LB table rows...
WhereCache(func (lb *LoadBalancer) bool {
    return lb.ExternalIds["myIdKey"] == "myIdValue"
}).List(ctx, &results)
```

Now you can have
```
dbModel, err := nbdb.FullDatabaseModel()
db.SetIndexes(map[string][]model.ClientIndex{
  "Load_Balancer": {{Columns: []model.ColumnKey{Column: "external_ids", Key: "myIdKey"}}}
})

// connect and everything....

lb := &LoadBalancer{
    ExternalIds: map[string]string{"myIdKey": "myIdValue"},
}
// quick indexed result
Where(lb).List(ctx, &results)
```

Other changes in this context:

* Cherry-picked @dave-tucker's [api: Allow use of Where for client-side ops](https://github.com/ovn-org/libovsdb/commit/b54c97c4b95409f779814be465a21f252082554a) 
  so that client indexes can be leveraged through the `Where.List` and `WhereAll.List` 
  conditional APIs. 
* As improvements to the above, run the predicates on shallow copies of model, 
  remove the unnecessary clone and return cache matches directly instead of 
  in-directing through RowsByCondition.
* Fixed an issue with RowsByCondition that was implementing an `Or` between the 
  conditions instead of an `And`. also improved performance.
* Instead of translating model conditions to ovsdb conditions and
  sending those to the server, perform a query on the cache through
  `RowsByModel` or `RowsByCondition` and generate uuid based conditions
  off those results. Fallback to the original conditionals if no cache hits are obtained.

  This is similar to what was already being done for the predicate conditional, this PR
  does the same for the equality and explicit conditionals.

  This has two benefits:

  * OVSDB server does not index schema indexes, it only ensures uniqueness
    but operations based on those indexes are not necessarily fast. By
    querying the cache for the uuids we improve the performance of index
    based operations. Users will no longer be required to do a `Get` before any 
    server side op to ensure this is optimized through the uuid.
  * The conditionals APIs can fully benefit of client indexes for both cache and 
    server side ops.
  
**RFCs**

* The `ClientDBModel`API could be better but I did not want to complicate it 
  unnecessarily. Feedback welcome.
* With this change we rely more on the cache. But the cache doesn't seem to
  be aware at all of monitored vs non-monitored columns and will happily index
  zero values from non-monitored columns. I am not worried in excess about it
  in the context of this PR as this is an issue already present, but perhaps we
  ought to do something about it.

jcaamano commented:
Added a couple of commits to benchmark & optimize `RowsByCondition` so that it makes use of indexes as much as possible.

Results before this PR:
```
goos: linux
goarch: amd64
pkg: github.com/ovn-org/libovsdb/cache
cpu: Intel Core Processor (Skylake, IBRS)
BenchmarkRowsByCondition/by_uuid-4         	      38	  28737606 ns/op	 6850409 B/op	  140222 allocs/op
BenchmarkRowsByCondition/by_single_column_squema_index-4         	    1686	    742480 ns/op	     925 B/op	      16 allocs/op
BenchmarkRowsByCondition/by_multi_column_squema_index-4          	      19	  60704196 ns/op	13540612 B/op	  280450 allocs/op
BenchmarkRowsByCondition/by_two_squema_indexes-4                 	     817	   1537409 ns/op	    1516 B/op	      29 allocs/op
BenchmarkRowsByCondition/by_squema_index_and_non-index-4         	      37	  31416413 ns/op	 6772972 B/op	  140246 allocs/op
BenchmarkRowsByCondition/by_single_non_index-4                   	      37	  30643479 ns/op	 6770317 B/op	  140226 allocs/op
BenchmarkRowsByCondition/by_multiple_non_indexes-4               	      19	  60734832 ns/op	13542517 B/op	  280456 allocs/op
PASS
ok  	github.com/ovn-org/libovsdb/cache	8.821s
```

Results before optimization commit (includes using single-column indexes, match all logic and avoids unnecessary clones):
```
goos: linux
goarch: amd64
pkg: github.com/ovn-org/libovsdb/cache
cpu: Intel Core Processor (Skylake, IBRS)
BenchmarkRowsByCondition/by_uuid-4         	  272794	      4057 ns/op	    1181 B/op	      20 allocs/op
BenchmarkRowsByCondition/by_single_column_squema_index-4         	  285651	      4229 ns/op	    1181 B/op	      20 allocs/op
BenchmarkRowsByCondition/by_single_column_client_index-4         	  288646	      4283 ns/op	    1181 B/op	      20 allocs/op
BenchmarkRowsByCondition/by_multi_column_squema_index-4          	     400	   2986806 ns/op	  241852 B/op	   20036 allocs/op
BenchmarkRowsByCondition/by_two_squema_indexes-4                 	  224726	      5065 ns/op	    1645 B/op	      26 allocs/op
BenchmarkRowsByCondition/by_squema_index_and_non-index-4         	  201648	      5581 ns/op	    1749 B/op	      32 allocs/op
BenchmarkRowsByCondition/by_single_non_index-4                   	     387	   3047246 ns/op	  241289 B/op	   20024 allocs/op
BenchmarkRowsByCondition/by_multiple_non_indexes-4               	     392	   2976273 ns/op	  241866 B/op	   20036 allocs/op
PASS
ok  	github.com/ovn-org/libovsdb/cache	11.000s
```

Results after optimization (includes using multi-column indexes):
```
goos: linux
goarch: amd64
pkg: github.com/ovn-org/libovsdb/cache
cpu: Intel Core Processor (Skylake, IBRS)
BenchmarkRowsByCondition/by_uuid-4         	  269812	      4106 ns/op	    1221 B/op	      22 allocs/op
BenchmarkRowsByCondition/by_single_column_squema_index-4         	  198980	      5833 ns/op	    1797 B/op	      36 allocs/op
BenchmarkRowsByCondition/by_single_column_client_index-4         	  178773	      6490 ns/op	    2013 B/op	      45 allocs/op
BenchmarkRowsByCondition/by_multi_column_squema_index-4          	   92376	     12630 ns/op	    5090 B/op	     112 allocs/op
BenchmarkRowsByCondition/by_two_squema_indexes-4                 	  109209	     10330 ns/op	    3957 B/op	      88 allocs/op
BenchmarkRowsByCondition/by_squema_index_and_non-index-4         	  112953	     10330 ns/op	    3957 B/op	      93 allocs/op
BenchmarkRowsByCondition/by_single_non_index-4                   	     397	   3073535 ns/op	  241862 B/op	   20042 allocs/op
BenchmarkRowsByCondition/by_multiple_non_indexes-4               	     391	   2958540 ns/op	  244022 B/op	   20099 allocs/op
PASS
ok  	github.com/ovn-org/libovsdb/cache	10.820s
```

Being able to use multi-column indexes does have a cost for the single-column index performance but I think it makes sense to have it. It's an overall good improvement for a start and I am sure this can be optimized further in various ways. Ideas welcome though if you see easy picks.

dcbw commented on client/api.go:
```diff
@@ -166,8 +166,7 @@ func (a api) List(ctx context.Context, result interface{}) error {
 			}
-			m := model.Clone(row)
-			appendValue(reflect.ValueOf(m))
+			appendValue(reflect.ValueOf(row))
```
@jcaamano it's worth explaining in the commit message for this specific commit that the Clone() is unecessary because RowsByCondition() already clones the returned rows.

dcbw commented on cache/cache.go:
```diff
@@ -113,30 +113,31 @@ func (r *RowCache) Row(uuid string) model.Model {
 }
 
 // RowByModel searches the cache using a the indexes for a provided model
```
@jcaamano can you update the code doc here too? `using a the` -> `using the` and also add something about it returning the UUID of the found row.

jcaamano commented:
Pushed an update that shortcuts RowsByCondition when we have reduced the possible index matches to 1. That will help in most situations where I would think we would expect a single match. As said, this can be tuned and optimized further but don't want to go overboard with no experience.

Other than that just added some additional tests and improvement to godoc  that come to mind as times goes on.

Updated results, added a benchmark that uses many conditions.

Results before this PR:
```
goos: linux
goarch: amd64
pkg: github.com/ovn-org/libovsdb/cache
cpu: Intel Core Processor (Skylake, IBRS)
BenchmarkRowsByCondition/by_uuid-4         	      39	  31181645 ns/op	 7156346 B/op	  140223 allocs/op
BenchmarkRowsByCondition/by_single_column_squema_index-4         	    1608	    719762 ns/op	     956 B/op	      16 allocs/op
BenchmarkRowsByCondition/by_multi_column_squema_index-4          	      16	  65336847 ns/op	14151228 B/op	  280448 allocs/op
BenchmarkRowsByCondition/by_two_squema_indexes-4                 	     800	   1582913 ns/op	    1577 B/op	      29 allocs/op
BenchmarkRowsByCondition/by_squema_index_and_non-index-4         	      33	  33603397 ns/op	 7077108 B/op	  140241 allocs/op
BenchmarkRowsByCondition/by_single_non_index-4                   	      32	  32456669 ns/op	 7076734 B/op	  140228 allocs/op
BenchmarkRowsByCondition/by_multiple_non_indexes-4               	      18	  64228400 ns/op	14151768 B/op	  280449 allocs/op
BenchmarkRowsByCondition/by_many_conditions-4                    	       8	 141383613 ns/op	33114615 B/op	  600961 allocs/op
PASS
ok  	github.com/ovn-org/libovsdb/cache	9.916s
```

Results before optimization commit (includes using single-column indexes, match all logic and avoids unnecessary clones):
```
goos: linux
goarch: amd64
pkg: github.com/ovn-org/libovsdb/cache
cpu: Intel Core Processor (Skylake, IBRS)
BenchmarkRowsByCondition/by_uuid-4         	  207000	      5366 ns/op	    1854 B/op	      32 allocs/op
BenchmarkRowsByCondition/by_single_column_squema_index-4         	  215368	      5775 ns/op	    1854 B/op	      32 allocs/op
BenchmarkRowsByCondition/by_single_column_client_index-4         	     166	   6692064 ns/op	 1842801 B/op	   50041 allocs/op
BenchmarkRowsByCondition/by_multi_column_squema_index-4          	     370	   2896103 ns/op	  242527 B/op	   20048 allocs/op
BenchmarkRowsByCondition/by_two_squema_indexes-4                 	  161007	      6484 ns/op	    2318 B/op	      38 allocs/op
BenchmarkRowsByCondition/by_squema_index_and_non-index-4         	  154776	      6872 ns/op	    2428 B/op	      44 allocs/op
BenchmarkRowsByCondition/by_single_non_index-4                   	     387	   3063313 ns/op	  241962 B/op	   20036 allocs/op
BenchmarkRowsByCondition/by_multiple_non_indexes-4               	     396	   2947208 ns/op	  242533 B/op	   20048 allocs/op
BenchmarkRowsByCondition/by_many_conditions-4                    	   89371	     12340 ns/op	    5422 B/op	      93 allocs/op
PASS
ok  	github.com/ovn-org/libovsdb/cache	13.949s
```

Results after optimization (includes using multi-column indexes):
```
goos: linux
goarch: amd64
pkg: github.com/ovn-org/libovsdb/cache
cpu: Intel Core Processor (Skylake, IBRS)
BenchmarkRowsByCondition/by_uuid-4         	  208352	      5577 ns/op	    1894 B/op	      34 allocs/op
BenchmarkRowsByCondition/by_single_column_squema_index-4         	  160498	      7098 ns/op	    2428 B/op	      47 allocs/op
BenchmarkRowsByCondition/by_single_column_client_index-4         	  111760	      9768 ns/op	    3677 B/op	      75 allocs/op
BenchmarkRowsByCondition/by_multi_column_squema_index-4          	   79140	     14264 ns/op	    5667 B/op	     123 allocs/op
BenchmarkRowsByCondition/by_two_squema_indexes-4                 	  149360	      7876 ns/op	    2910 B/op	      54 allocs/op
BenchmarkRowsByCondition/by_squema_index_and_non-index-4         	  128370	      7967 ns/op	    2910 B/op	      54 allocs/op
BenchmarkRowsByCondition/by_single_non_index-4                   	     390	   3027351 ns/op	  242536 B/op	   20054 allocs/op
BenchmarkRowsByCondition/by_multiple_non_indexes-4               	     403	   2975605 ns/op	  244693 B/op	   20111 allocs/op
BenchmarkRowsByCondition/by_many_conditions-4                    	   95650	     12171 ns/op	    5694 B/op	      89 allocs/op
PASS
ok  	github.com/ovn-org/libovsdb/cache	12.210s
```

Was curious to have a benchmark of index vs predicate. It turns out there was one for predicates but I don't it was working fine (too good to be true) so I updated it. Results: 
```
BenchmarkAPIListPredicate/ApiListPredicate:_predicate_returns_none-4         	     363	   3117351 ns/op	  930387 B/op	   20010 allocs/op
BenchmarkAPIListPredicate/ApiListPredicate:_predicate_returns_all-4          	     225	   5167648 ns/op	 2445557 B/op	   30223 allocs/op
BenchmarkAPIListPredicate/ApiListPredicate:_predicate_on_an_arbitrary_condition-4         	     332	   3627668 ns/op	 1123775 B/op	   21152 allocs/op
BenchmarkAPIListPredicate/ApiListPredicate:_predicate_matches_name-4                      	     363	   3292371 ns/op	  930691 B/op	   20012 allocs/op
BenchmarkAPIListPredicate/ApiListPredicate:_by_index,_no_predicate-4                      	   45225	     25121 ns/op	   84076 B/op	      32 allocs/op
```

jcaamano commented:
Would `ColumnKey` be better served typed simpler as `map[string][]interface{}` as a map of columns to optional keys. 
Or at least 
```
struct {
  Column string
  Keys []interface{}
}
````
Really nothing prevents using multiple keys within the same map column and thus might be cleaner that way.

flavio-fernandes commented:
SUPER NIT: fix typo in commit description of commit
https://github.com/ovn-org/libovsdb/pull/306/commits/0d2fa37e8d4fb3d291768aaac80e32c4c66a85da
```
Do use RowsByCondition to add support to match against the cache for the
 --------------------------------------------------------^
explicit conditional.
```

flavio-fernandes commented on cache/cache.go:
```diff
@@ -261,18 +352,25 @@ func (r *RowCache) IndexExists(row model.Model) error {
 		}
-		if existing, ok := vals[val]; ok && existing != uuid.(string) {
+		vals := r.indexes[index]
+		if existing := vals[val]; !existing.empty() && existing.getAny() != uuid.(string) {
```
!!

flavio-fernandes replied:
Ah, I get it now... when using schemaIndex, this set has at most one element.
May be worth putting a comment about that here?

jcaamano replied:
same as before

flavio-fernandes commented on cache/cache.go:
```diff
@@ -742,29 +863,49 @@ func (t *TableCache) Run(stopCh <-chan struct{}) {
+	for _, clientIndex := range clientIndexes {
+		columnKeys := clientIndex.Columns
+		index := newIndexFromColumnKeys(columnKeys...)
+		// this is already a DB index, ignore
```
NIT: if this is...

flavio-fernandes commented on client/condition.go:
```diff
@@ -82,7 +82,12 @@ func (c *equalityConditional) Matches() (map[string]model.Model, error) {
 // Generate returns a condition based on the model and the field pointers
 func (c *equalityConditional) Generate() ([][]ovsdb.Condition, error) {
-	return generateConditionsFromModels(c.cache.DatabaseModel(), map[string]model.Model{"": c.model})
+	models, _ := c.Matches()
```
hmm.... I like it but, I fear this means very different things depending on what is in the cache at the time.
This will blurry the lines a bit on the line drawn in regards to where the condition is executed, yes?
Maybe that is the intention here.
A less magic approach would be to have `Generate()` and `GenerateFromCache()` flavors? Just a thought.

jcaamano replied:
Well, the counter arguments to your comment are:

1. It makes sense that rows modified with `Where(...).Update/Delete(...)` would be the same as those obtained from the equivalent `Where(...).List()` using the  same conditions at the same time.
2. For the most part, operating on non UUID conditions on the server is inefficient. In OVN-K, we already do `Get` before any update to get the UUID, and that's one of, if not, the main use of having a cache. If you don't have a cache, this would change nothing as it would use the original conditions.
3. If the condition is executed at the server, then it also depends on what's in the server at that point in time as well, right? Meaning, a client user must already be expecting and prepared for inconsistent results and have mechanisms in place to prevent or fix that if overlapping transactions. If there are no overlapping transactions, this is a non issue. 
A client user does not know what's the state of the server data just as it does not know what's the state of the cache data unless it's the only client and transactions are sequenced and then it doesn't really matter. So overall, I don't think this changes much in regards to what a client user is in control of here.

I don't having two methods makes sense, it would be one or the other. Unless we add new APIs.

flavio-fernandes replied:
> Well, the counter arguments to your comment are

fair enougn. @jcaamano . I think that if we did anything, it would be on the API. But let's not worry about that at this point. Thank you for sharing your thoughts on this.

flavio-fernandes commented on client/condition.go:
```diff
@@ -126,7 +131,12 @@ func (c *explicitConditional) Matches() (map[string]model.Model, error) {
 // Generate returns a condition based on the model and the field pointers
 func (c *explicitConditional) Generate() ([][]ovsdb.Condition, error) {
-	return c.anyConditions, nil
+	models, _ := c.Matches()
```
!!

flavio-fernandes commented on cache/cache.go:
```diff
@@ -218,34 +306,37 @@ func (r *RowCache) Update(uuid string, m model.Model, checkIndexes bool) error {
 		// check that there are no conflicts
-		if conflict, ok := vals[newVal]; ok && checkIndexes && conflict != uuid {
+		vals := r.indexes[index]
+		if existing, ok := vals[newVal]; ok && indexSpec.isSchemaIndex() && checkIndexes && existing.getAny() != uuid {
```
Ah, I get it now... when using schemaIndex, this set has at most one element.
May be worth putting a comment about that here?

jcaamano replied:
Will replace with `!existing.has(uuid)` which I think makes more sense.

flavio-fernandes reviewed (commented):
some initial thoughts and reactions.
But... I think this is great!

flavio-fernandes commented:
@amorenoz @dave-tucker PTAL as well!

flavio-fernandes commented on cache/cache_test.go:
```diff
@@ -2567,6 +2575,7 @@ func TestTableCacheRowsByCondition(t *testing.T) {
 		{
 			"by excludes uuid, multiple results",
 			[]ovsdb.Condition{{Column: "_uuid", Function: ovsdb.ConditionNotEqual, Value: ovsdb.UUID{GoUUID: "foo"}}},
```
hm.... `ovsdb.ConditionExcludes` ?

jcaamano replied:
Why?

flavio-fernandes replied:
> Why?

My bad for not being clear enough. I think that in line 2569 / 2577 we should use:

From:
`[]ovsdb.Condition{{Column: "_uuid", Function: ovsdb.ConditionNotEqual, Value: ovsdb.UUID{GoUUID: "foo"}}},`
To:
`[]ovsdb.Condition{{Column: "_uuid", Function: ovsdb.ConditionExcludes, Value: ovsdb.UUID{GoUUID: "foo"}}},`

jcaamano replied:
Ah! I didn't get it. Sorry!

flavio-fernandes commented on cache/cache_test.go:
```diff
@@ -1563,3 +2315,558 @@ func TestTableCacheApplyModifications(t *testing.T) {
+					"max": "unlimited"
+				}
+			  },
+			  "foobar": {
```
the duplication of foobar column... was it on purpose?

jcaamano replied:
nope :P

flavio-fernandes reviewed (changes requested):
I have just a silly nit on the duplication of foobar column in the test.
otherwise, lgtm @jcaamano !

dcbw commented:
@jcaamano DCO check failure?

jcaamano commented:
> @jcaamano DCO check failure?
> 
> EDIT: ah your fixup. I assume you'll rebase + squash before merge

rebased & squashed

I am going to run this by OVN-K to make sure everything looks fine. 

I would also like feedback from @dave-tucker and @amorenoz at least on the API changes and the RFCs.

amorenoz commented on cache/cache.go:
```diff
@@ -438,40 +446,217 @@ func (r *RowCache) RowsShallow() map[string]model.Model {
+// Note that this method does not guarantee that the models of those uuids match
+// all the provided conditions. Thus the caller is required to evaluate all the
+// conditions against those models. This is only useful to trim down the
+// probable matches of models fast via indexes.
```
Maybe a more explanatory commit message that outlines the high level optimization strategy would be useful to understand this.

flavio-fernandes commented on cache/cache.go:
```diff
@@ -438,40 +446,217 @@ func (r *RowCache) RowsShallow() map[string]model.Model {
+// all the provided conditions. Thus the caller is required to evaluate all the
+// conditions against those models. This is only useful to trim down the
+// probable matches of models fast via indexes.
+//nolint:gocyclo // warns overall function is complex but ignores anonymous functions
```
:) +1

jcaamano replied:
> OK, I'm not an expert in this code and I am rusty on the project itself but having read this function 3 times I agree with gocyclo: This function (plus the next one) is complex.

The function has above average complexity alright. The point of the comment is that if those inner functions where instead outer functions, what would be basically an indentation change only for the most part, the complexity would be the same and go-lint would not complain.

> Maybe moving all this index-based condition matching algorithm to it's own file with it's own "public" type, internal helper types (e.g: indexableCondition), public functions and (broken-down) helper functions and UT would make it easier to digest and maintain.

Yes, this could be laid out better. I decided to keep it as is because it is a definitive improvement, pretty ad-hoc, not too long, and testing coverage is good.

This is non trivial effort though, so what I can suggest is that we pull out the `RowsByCondition` optimization commit to a separate PR as to not stale the remaining effort on this PR.

jcaamano commented on client/api.go:
```diff
@@ -147,23 +147,29 @@ func (a api) List(ctx context.Context, result interface{}) error {
+		}
+		for _, row := range matchingRows {
+			if i >= resultVal.Cap() {
+				break
```
Yeah, although this was a direct cherry pick from @dave-tucker PR. I will just fix and squash all the relevant fixes to this PR.

jcaamano commented on client/api.go:
```diff
@@ -147,23 +147,29 @@ func (a api) List(ctx context.Context, result interface{}) error {
-			break
+	var conditions [][]ovsdb.Condition
+	if a.cond != nil {
+		conditions, err = a.cond.Generate()
```
I will update docs as relevant. 

But I am unsure which specific doc update would apply to this portion of the code. Other that godoc that should be automatically generated, nothing comes to mind?

jcaamano commented on client/condition.go:
```diff
@@ -13,121 +13,140 @@ import (
 	if err != nil {
 		return nil, err
 	}
+	anyConditions, err := generateOvsdbConditionsFromModelConditions(cache.DatabaseModel(), info, cond, all)
```
It must, because that is where it is coming from in the first place:
https://github.com/ovn-org/libovsdb/blob/3f40c8abb6ff6ecddb3a151908c956e50365219d/client/api.go#L202-L224

I think there is no need to pass the model in a separate argument.

jcaamano commented:
Pushed an update:
* Addressed all the comments up until now
* Moved commit [Optimize RowsByCondition](https://github.com/ovn-org/libovsdb/pull/306/commits/f6b3d1785961c340d49b939b11f6fef65381e8f6) to the end, to optionally remove from this PR.
* Updated docs & README
* Reviewing the docs, I saw that there was a concern sending to the server a high number of ops resulting from a `WhereCache` evaluation. Evaluating explicit conditions against the cache before sending them to the server leads to the same concern. Thus I split that into a separate commit to optionally remove it from this  PR as well.

jcaamano commented on cache/cache.go:
```diff
@@ -1053,9 +1194,37 @@ func valueFromIndex(info *mapper.Info, index index) (interface{}, error) {
 		val := hex.EncodeToString(h.Sum(buf.Bytes()))
```
how so? It's github maybe using different spacing for a tab?

jcaamano commented:
Testing with ovn-k not showing any evident issues
https://github.com/ovn-org/ovn-kubernetes/pull/2989

flavio-fernandes commented:
/lgtm

dave-tucker reviewed (approved):
@jcaamano this is excellent work, thank you very much!
While I'm context switching (from Rust and eBPF) the comments and docs helped a lot when reviewing this... and there are tests too. I don't have anything to add that hasn't been covered by other reviewers, other than adding my LGTM :sunglasses:

amorenoz commented:
Sorry, I was preempted and couldn't finish the review on time. Thanks @jcaamano for the great work!

dave-tucker commented:
Ack. I'm travelling this week but might be able to look on Weds.

On Tue, 17 May 2022, 10:26 Jaime Caamaño Ruiz, ***@***.***>
wrote:

> @jcaamano <https://github.com/jcaamano> DCO check failure?
>
> EDIT: ah your fixup. I assume you'll rebase + squash before merge
>
> rebased & squashed
>
> I am going to run this by OVN-K to make sure everything looks fine.
>
> I would also like feedback from @dave-tucker
> <https://github.com/dave-tucker> and @amorenoz
> <https://github.com/amorenoz> at least on the API changes and the RFCs.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ovn-org/libovsdb/pull/306#issuecomment-1128635700>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA43JEPTE2QSCGM56KZQYQTVKNQ5NANCNFSM5VQ7RMCA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
